Fix karma, readme updates, combine source to one file

ryanio opened the pull request:
This PR updates the karma-typescript dependency to a built version from master using gitpkg for the unreleased [fix](https://github.com/monounity/karma-typescript/pull/502).

It also includes some refactoring and readme updates.

ryanio committed:
update karma-typescript to use fix on master not yet released: https://github.com/monounity/karma-typescript/pull/502

ryanio committed:
update readme

ryanio committed:
bin/rlp: use capital RLP, destructure bytesToHex

ryanio committed:
destructure ternary operator for easier readability

ryanio committed:
tidy, touch ups, flatten return statements when possible, move utils to own file

ryanio committed:
destructure utils usage, move numberToBytes to test utils

ryanio committed:
eslint updates

ryanio committed:
rename baToJSON to arrToJSON

paulmillr commented:
I know using 3 files instead of 1 is much more readable, just not sure about ESM support - or Deno support. 

I have been making my stuff work with those, and there are some things that you need to take into account. 

For example, in Deno you can't import .js files, and you need to specify extensions in imports. In ESM, you cannot import .ts files.

My suggestion would be to move everything into one file and place it in root. After that, after the release, we could work on .esm tsconfig and Deno package. I could help here

paulmillr commented:
I have worked around "many files" thing in bls12-381 by using export maps but we definitely don't want to force Deno users to do that (a separate .json file)

ryanio committed:
add new test coverage from @jochem-brouwer, add es2020 property to eslint, renamings

ryanio committed:
update package.json version to 3.0.0, add changelog entries

ryanio committed:
remove separate browser build since tsconfigs are the same. tested with browserify and was able to RLP encode in the browser from dist/index.js

ryanio committed:
fix karma-typescript

ryanio committed:
fix karma-typescript - another attempt

ryanio committed:
ci:
  - use npm 7 for github url integrity support (for karma-typescript unreleased workaround)
  - run coverage inside node v16 workflow so a duplicative job is not needed

ryanio commented:
@paulmillr sounds good, thanks for the feedback, we already had 2 files (index, types) so I thought a third for utils would be harmless (the improved readability is great), but I am also in favor of broader ecosystem support, so whatever we need to do here - if it is one file then that's fine by me too.

Appreciate your help with esm and Deno insights, we can probably bring these learnings over to our monorepo as well.

ryanio committed:
ci: more clarity on reason for added npm v7 and when it can be removed

ryanio commented:
For the karma-typescript fix, I wrote some of my findings in this comment: https://github.com/ethereumjs/ethereumjs-monorepo/pull/1602#issuecomment-1008222864

ryanio committed:
combine to one file to improve esm, deno support

ryanio commented:
@talentlessguy has helped us with Deno compatibility in the past, let us know how this PR is looking

ryanio committed:
update changelog

ryanio committed:
Add NestedUint8Array type (fixes issue #78)

ryanio committed:
tidy readme

ryanio committed:
changelog: add cli PR

ryanio committed:
nit: keys by alphabetical order

paulmillr commented on .github/workflows/build.yml:
```diff
@@ -12,7 +12,7 @@ jobs:
     strategy:
       matrix:
-        node-version: [12.x, 13.x, 14.x, 16.x]
+        node-version: [12, 13, 14, 16]
```
Let's remove v13, it has not been supported for many years.

paulmillr reviewed (changes requested):
Looks great to me!

ryanio committed:
replace node 13 with 17

ryanio commented:
thanks!
